colcontainer: don't call writeFooterAndFlush in Close after an error

asubiotto opened the pull request:
When disk monitoring was added to the diskQueue, internal state was left
unmodified in case of an error. This would cause the diskQueue to attempt to
write to disk again on Close even if an out of disk error occured during normal
operation. This commit sets the serializer to nil after Finish has been called
so that if a subsequent call to Close happens, the diskQueue does not try to
write footer information (these files will be deleted anyway later on). When
an error doesn't happen, the diskQueue expects the serializer to be reset.

Release note (bug fix): unexpected internal errors containing stack traces that
reference a countingWriter nil pointer have now been fixed.

Closes #59421 
Closes #54552

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59477)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59477#-:-MS4QQ131dXpDObqfjQf:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

Commit:
colcontainer: don't call writeFooterAndFlush in Close after an error

When disk monitoring was added to the diskQueue, internal state was left
unmodified in case of an error. This would cause the diskQueue to attempt to
write to disk again on Close even if an out of disk error occured during normal
operation. This commit sets the serializer to nil after Finish has been called
so that if a subsequent call to Close happens, the diskQueue does not try to
write footer information (these files will be deleted anyway later on). When
an error doesn't happen, the diskQueue expects the serializer to be reset.

Release note (bug fix): unexpected internal errors containing stack traces that
reference a countingWriter nil pointer have now been fixed.

asubiotto commented:
TFTR 

bors r=yuzefovich
